client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,1695 @@
+#[derive(Debug)]
+pub struct ProtocolController {
+	/// Set id to use when sending connect/drop requests to `Notifications`.
+	// Will likely be replaced by `ProtocolName` in the future.
+	set_id: SetId,
+	/// Receiver for outer API messages from [`ProtocolHandle`].
+	actions_rx: TracingUnboundedReceiver<Action>,
+	/// Receiver for connection events from `Notifications` sent via [`ProtocolHandle`].
+	events_rx: TracingUnboundedReceiver<Event>,
+	/// Number of occupied slots for incoming connections (not counting reserved nodes).
+	num_in: u32,
+	/// Number of occupied slots for outgoing connections (not counting reserved nodes).
+	num_out: u32,
+	/// Maximum number of slots for incoming connections (not counting reserved nodes).
+	max_in: u32,
+	/// Maximum number of slots for outgoing connections (not counting reserved nodes).
+	max_out: u32,
+	/// Connected regular nodes.
+	nodes: HashMap<PeerId, Direction>,
+	/// Reserved nodes. Should be always connected and do not occupy peer slots.
+	reserved_nodes: HashMap<PeerId, PeerState>,
+	/// Connect only to reserved nodes.
+	reserved_only: bool,
+	/// Next time to allocate slots. This is done once per second.
+	next_periodic_alloc_slots: Instant,
+	/// Outgoing channel for messages to `Notifications`.
+	to_notifications: TracingUnboundedSender<Message>,
+	/// `PeerStore` handle for checking peer reputation values and getting connection candidates
+	/// with highest reputation.
+	peer_store: Box<dyn PeerReputationProvider>,
+}
+
+impl ProtocolController {
+	/// Construct new [`ProtocolController`].
+	pub fn new(
+		set_id: SetId,
+		config: SetConfig,
+		to_notifications: TracingUnboundedSender<Message>,
+		peer_store: Box<dyn PeerReputationProvider>,
+	) -> (ProtocolHandle, ProtocolController) {
+		let (actions_tx, actions_rx) = tracing_unbounded("mpsc_api_protocol", 10_000);
+		let (events_tx, events_rx) = tracing_unbounded("mpsc_notifications_protocol", 10_000);
+		let handle = ProtocolHandle { actions_tx, events_tx };
+		let reserved_nodes =
+			config.reserved_nodes.iter().map(|p| (*p, PeerState::NotConnected)).collect();
+		let controller = ProtocolController {
+			set_id,
+			actions_rx,
+			events_rx,
+			num_in: 0,
+			num_out: 0,
+			max_in: config.in_peers,
+			max_out: config.out_peers,
+			nodes: HashMap::new(),
+			reserved_nodes,
+			reserved_only: config.reserved_only,
+			next_periodic_alloc_slots: Instant::now(),
+			to_notifications,
+			peer_store,
+		};
+		(handle, controller)
+	}
+
+	/// Drive [`ProtocolController`]. This function returns when all instances of
+	/// [`ProtocolHandle`] are dropped.
+	pub async fn run(mut self) {
+		while self.next_action().await {}
+	}
+
+	/// Perform one action. Returns `true` if it should be called again.
+	///
+	/// Intended for tests only. Use `run` for driving [`ProtocolController`].
+	pub async fn next_action(&mut self) -> bool {
+		// Perform tasks prioritizing connection events processing.
+		let either = loop {
```

Comment:
I have to say I don't like this at all. Why do we still have to prioritize events if the `IncomingIndex` is used for verifying whether a command should be processed/ignored?

Also what happens if the prioritization breaks? Let's say in the `futures::select!` we get an action from `action_rx` and then a microsecond later `events_rx` would yield an event. Is that going to cause an issue because now there is an event pending? Is there some threshold for how many accidental and incorrect prioritizations we can tolerate before something breaks?  This is my problem with this because looking at this code, it's not clear what happens if the prioritization breaks and what are the effects of it.

If there's no way to get rid of this, is it possible to use `futures::select_biased!` and add a comment what happens if the prioritization is broken?

Reply:
I don't like it either, but for fuzz test to pass I needed both to drop obsolete commands before receiving the answer for incoming connection AND prioritize the events processing from `Notifications` over other actions. I don't think that we are completely bulletproof from troubles with prioritization, but at least the fuzz test didn't reveal any issues.

I would like to have a proper solution to this problem instead, but I don't have one.

Thanks for pointing out `futures::select_biased!`, this is what I was looking for but wasn't able to find.

I'm going to run fuzz test again without prioritization to recall specific issues and add the comment.

Reply:
Alright sad to hear but that's the reality I guess. The comment would be good because if someone some day decided to refactor `ProtocolController` and there is no explanation for the prioritization, it can be hard to understand why the code is written the way it is.

Reply:
I have added a detailed comment to the module documentation on the way we try to keep `ProtocolController`'s peer states view consistent.